### PR TITLE
Web Share demo using GET for text and POST for images

### PR DESCRIPTION
**Pull request by PEConn**
Hey Andre and Pavol,

When I was looking at Web Share I put this demo together, but didn't get around to polishing it up or landing it. Maybe it will be useful for you to figure out how to get stuff working.

Peter C

### CODE REVIEW

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (padr31) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java:1]: Other places as well potentially

File: demos/twa-web-share-target/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            <!-- End of Web Share Target Changes. -->
+
+            <intent-filter android:autoVerify="true">
+                <action android:name="android.intent.action.VIEW" />
+
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.BROWSABLE" />
+
+                <data
+                    android:host="peconn.github.io"
```
Comment (padr31) [inline demos/twa-web-share-target/src/main/AndroidManifest.xml:49]: The page does not seem to be a PWA, not sure if that is intended

File: demos/twa-web-share-target/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            android:exported="true">
+            <intent-filter>
+                <action android:name="android.support.customtabs.trusted.TRUSTED_WEB_ACTIVITY_SERVICE" />
+
+                <category android:name="android.intent.category.DEFAULT" />
+            </intent-filter>
+        </service>
+    </application>
+
+</manifest>
```
Comment (padr31) [inline demos/twa-web-share-target/src/main/AndroidManifest.xml:111]: NIT: missing newline at eof

Comment (padr31) [reply demos/twa-web-share-target/src/main/AndroidManifest.xml:111]: At other places as well like MainActivity

File: settings.gradle [hunk 4]
```diff
@@ -1,5 +1,11 @@
-include ':demos:twa-custom-launcher', ':androidbrowserhelper', ':demos:twa-multi-domain', ':playbilling'
-include ':demos:twa-basic', ':demos:twa-webview-fallback'
+include ':androidbrowserhelper'
+include ':playbilling'
+
```
Comment (padr31) [inline settings.gradle:3]: NIT: remove new line

**Review by padr31 — APPROVED** [review]
Looks good, added mostly NITs

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreban) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java]: done

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * Creates a {@link ShareData} object from an {@link Intent}. Returns null if the intent is not
+     * a share intent, i.e. its action isn't {@link Intent#ACTION_SEND} or
+     * {@link Intent#ACTION_SEND_MULTIPLE}.
+     */
+    @Nullable
+    public static ShareData retrieveShareDataFromIntent(Intent intent) {
+        String action = intent.getAction();
+        if (!Intent.ACTION_SEND.equals(action) && !Intent.ACTION_SEND_MULTIPLE.equals(action)) {
```
[the remark this answers is not in the document]
Comment (andreban) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/SharingUtils.java]: done

File: demos/twa-web-share-target/src/main/AndroidManifest.xml (hunk 3 above)
Comment (andreban) [reply demos/twa-web-share-target/src/main/AndroidManifest.xml]: done

File: settings.gradle (hunk 4 above)
Comment (andreban) [reply settings.gradle]: done

File: demos/twa-web-share-target/src/main/AndroidManifest.xml (hunk 2 above)
Comment (andreban) [reply demos/twa-web-share-target/src/main/AndroidManifest.xml:61]: I think this is intentional for now. I'll replace when we have a better one to link to.

File: demos/twa-web-share-target/src/main/java/com/google/androidbrowserhelper/webshare/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+package com.google.androidbrowserhelper.webshare;
+
+import androidx.appcompat.app.AppCompatActivity;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.view.View;
+
+public class MainActivity extends AppCompatActivity {
```
Comment (PEConn) [inline demos/twa-web-share-target/src/main/java/com/google/androidbrowserhelper/webshare/MainActivity.java:23]: Could you make this extend Activity instead (and remove the dependency in build.gradle)?

File: demos/twa-web-share-target/src/main/AndroidManifest.xml (hunk 2 above)
Comment (PEConn) [reply demos/twa-web-share-target/src/main/AndroidManifest.xml:61]: https://peconn.github.io/starters is a PWA though

File: demos/twa-web-share-target/src/main/java/com/google/androidbrowserhelper/webshare/MainActivity.java (hunk 6 above)
Comment (andreban) [reply demos/twa-web-share-target/src/main/java/com/google/androidbrowserhelper/webshare/MainActivity.java]: done.

File: demos/twa-web-share-target/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -1,4 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
```
Comment (padr31) [inline demos/twa-web-share-target/src/main/AndroidManifest.xml:1]: Should this be below the copyright text?

**Review by padr31 — APPROVED** [review]

File: demos/twa-web-share-target/src/main/AndroidManifest.xml (hunk 7 above)
Comment (andreban) [reply demos/twa-web-share-target/src/main/AndroidManifest.xml:1]: According to the XML spec, the XML declaration must be the first line of the document, including comments. In practice, having the comments before that does indeed break builds.

**Comment by mvano** [comment]
Please take a look @PEConn @andreban so we can hopefully land this patch. If we find further issues we can fix them in follow up patches.

@flunderpero I agree with the changes you made in #83 - I found the same issues of an exception for incorrect use of getParcelableArrayListExtra and the necessity to call launchTrustedWebActivity when sharing a photo from e.g. Google Photos to the TWA. Otherwise access permission to the photo uri is not granted to the next activity. The actual fixes are simple and the same as what you found.

I have set up two sites to test this with, one using GET for sharing text and one using POST and a service worker for sharing photos:
https://mvano.github.io/share-text/index.html
https://mvano.github.io/share-files/index.html

These sites also use the Web Share API so we don't strictly need a native app to initiate a share action from. It also has digital asset link verification set up. Feel free to copy that and experiment with it as needed.

**Review by andreban — APPROVED** [review]
LGTM, thanks a lot for this!

**Comment by mvano** [comment]
@googlebot I consent

**Pull request merged**